// data-structure-typed/dist/data-structures/hash/coordinate-set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoordinateSet = void 0;
/**
 * data-structure-typed
 *
 * @author Tyler Zeng
 * @copyright Copyright (c) 2022 Tyler Zeng <zrwusa@gmail.com>
 * @license MIT License
 */
class CoordinateSet extends Set {
    constructor(joint) {
        super();
        this._joint = '_';
        if (joint !== undefined)
            this._joint = joint;
    }
    get joint() {
        return this._joint;
    }
    /**
     * The "has" function overrides the "has" method of the superclass and checks if a value exists in an array after
     * joining its elements with a specified separator.
     * @param {number[]} value - The parameter "value" is an array of numbers.
     * @returns The overridden `has` method is returning the result of calling the `has` method of the superclass, passing
     * in the joined value as an argument.
     */
    has(value) {
        return super.has(value.join(this._joint));
    }
    /**
     * The "add" function overrides the parent class's "add" function by joining the elements of the input array with a
     * specified delimiter before calling the parent class's "add" function.
     * @param {number[]} value - An array of numbers
     * @returns The overridden `add` method is returning the result of calling the `add` method of the superclass
     * (`super.add`) with the joined string representation of the `value` array (`value.join(this._joint)`).
     */
    add(value) {
        return super.add(value.join(this._joint));
    }
    /**
     * The function overrides the delete method and deletes an element from a Set by joining the elements of the input
     * array with a specified joint and then calling the delete method of the parent class.
     * @param {number[]} value - An array of numbers
     * @returns The `delete` method is returning the result of calling the `delete` method of the superclass, with the
     * `value` array joined together using the `_joint` property.
     */
    delete(value) {
        return super.delete(value.join(this._joint));
    }
    _setJoint(v) {
        this._joint = v;
    }
}
exports.CoordinateSet = CoordinateSet;
